### PR TITLE
Treat null as empty string for SyntaxTree's path

### LINKED ISSUE #12638: SyntaxTrees with a null path should be explicitly disallowed
At present, it is possible to construct a `SyntaxTree` with a `null` path in compiler that was compiled under the `Release` configuration. However, the documentation for `SyntaxTree.FilePath` explicitly states that `the path shall not be null` (http://source.roslyn.io/#Microsoft.CodeAnalysis/Syntax/SyntaxTree.cs,e12956b59d7683ea).

* CSharpParsedSyntaxTree (http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp/Syntax/CSharpSyntaxTree.ParsedSyntaxTree.cs,fcc0925d5e8c089c) has a `Debug.Assert` to ensure that no null path is used for the Debug configuration
* However, there is no such check in the public API `CSharpSyntaxTree.Create` (http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp/Syntax/CSharpSyntaxTree.cs,d40da3b7b4e39486). As such, it is only possible to construct a syntax tree with a null path in the Release configuration
* Note that the other entry points for creating a ParsedSyntaxTree do not allow a null path
 * `CSharpSyntaxTree.CreateWithoutClone` (http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp/Syntax/CSharpSyntaxTree.cs,ebf9a5453787d42b) always passes an empty path (`string.empty`)
 * `CSharpSyntaxTree.ParseText` (http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp/Syntax/CSharpSyntaxTree.cs,f4e9b1e7862d75c8) throws an exception for null path
 * `CSharpSyntaxTree.WithChanges` (http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp/Syntax/CSharpSyntaxTree.cs,f93c5917f27da6d2) uses the existing path
 * `CSharpSyntaxTree.WithFilePath` (http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp/Syntax/CSharpSyntaxTree.ParsedSyntaxTree.cs,ae8caa7afa30b1ab) passes the value thru to the constructor (only allowing null path in Release mode)
* VisualBasic matches the CSharp behavior

We should explicitly disallow null paths at the base layer by throwing an exception (rather than just asserting).

### PR DESCRIPTION
**Pull request by gafter**
The requirement was there previously in Debug (by assertion).
Now we convert a null into string.Empty.
Fixes #12638

### CODE REVIEW

**Comment by gafter** [comment]
Marking for personal review to let the tests run. If all is good I'll open it up for team review.

**Comment by gafter** [comment]
@tannergooding FYI

**Comment by gafter** [comment]
Failed a flaky test. See #13788. Testing again.

**Comment by tannergooding** [comment]
FYI. @JohnHamby, didn't we add some kind of workaround for this.

**Comment by JohnHamby** [comment]
This is a fine change. I don't remember where we coded around the possibility of a null path, but there is probably some dead code in the wake of this change.

**Comment by gafter** [comment]
@dotnet/roslyn-compiler @mattwar May I please have a couple of reviews?

**Comment by CyrusNajmabadi** [comment]
I would prefer we just drop the requirement.  From the IDE's perspective, using this field is just a bug at the Workspace layer or above.

File: src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxTreeTests.cs [hunk 1]
```diff
@@ -91,5 +92,16 @@ public void WithFilePath_DummyTree()
             Assert.Equal(newTree.FilePath, "new.cs");
             Assert.Equal(oldTree.ToString(), newTree.ToString());
         }
+
+        [Fact]
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxTreeTests.cs]: >Fact [](start = 9, length = 4)

WorkItem?

File: src/Compilers/VisualBasic/Test/Syntax/Syntax/SyntaxTreeTests.vb [hunk 2]
```diff
@@ -82,5 +82,15 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.UnitTests
             Assert.Equal(newTree.FilePath, "new.vb")
             Assert.Equal(oldTree.ToString(), newTree.ToString())
         End Sub
+
+        <Fact>
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Syntax/Syntax/SyntaxTreeTests.vb]: >Fact [](start = 9, length = 4)

WorkItem?

**Comment by AlekseyTs** [comment]
Changes under Compilers LGTM. It looks like there are no unit tests for changes under Workspaces.

**Comment by gafter** [comment]
> I would prefer we just drop the requirement.

That would actually be a larger change. There are assertions in many places that would have to be removed.

---
In reply to: [247386144](https://github.com/dotnet/roslyn/pull/13774#issuecomment-247386144) [](ancestors = 247386144)

**Comment by CyrusNajmabadi** [comment]
I would prefer it not for 'work size' reasons.  But because we actively do not believe this is a good API to have given the structure of the Workspace.  It is suitable for pure 'compilation API' only scenarios.  Which is what makes there be a tension around this API.

That said, I'd rather just relax things and consider any additional work to be bug level fixes.

**Comment by CyrusNajmabadi** [comment]
I'm not going to push hard on this though.  I'm just giving the perspective of IDE on the value of this feature.  Recall that in early discussions we didn't even want it on SyntaxTree to begin with.  If it's on SyntaxTree, we'd like it to be as loose a requirement as possible.

**Comment by gafter** [comment]
@mattwar What is your perspective on whether or not we should either enforce this restriction or remove it?

**Comment by mattwar** [comment]
I'm okay with adding the restriction. However, for backward compatibility sake, it would probably be better to just convert nulls to empty string in the constructor.

**Comment by gafter** [comment]
Revised to just convert nulls to empty string in the constructor. Reviewers, does this affect your opinion?

**Comment by AlekseyTs** [comment]
LGTM, modulo the failing tests.

**Comment by gafter** [comment]
@dotnet/roslyn-ide Can I please have a review from the IDE team?

File: src/Workspaces/Core/Portable/Workspace/Host/SyntaxTreeFactory/AbstractSyntaxTreeFactoryService.AbstractRecoverableSyntaxRoot.cs [hunk 3]
```diff
@@ -21,7 +21,7 @@ internal struct SyntaxTreeInfo
 
             public SyntaxTreeInfo(string filePath, ParseOptions options, ValueSource<TextAndVersion> textSource, Encoding encoding, int length)
             {
-                FilePath = filePath;
+                FilePath = filePath ?? string.Empty;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Host/SyntaxTreeFactory/AbstractSyntaxTreeFactoryService.AbstractRecoverableSyntaxRoot.cs]: Dumb question.  Why do we need this part?  If they asked for a null path, why not just keep the nll path?

Comment (gafter) [inline src/Workspaces/Core/Portable/Workspace/Host/SyntaxTreeFactory/AbstractSyntaxTreeFactoryService.AbstractRecoverableSyntaxRoot.cs]: I have no idea how many things in our own code base, and our customer code bases would have to be hardened to support null here. There are enough places that *prevented* that before that allowing it now is likely to break things.

**Comment by CyrusNajmabadi** [comment]
Hrmm...Not sure i get that logic.  If the null path was prevented before, then current existing code that produced null simply would not work.  Now it may not work, just elsewhere down the line.  That doesn't seem any better/worse than before.

That said, i honestly dont' care here.  So 👍 from me :)

**Pull request merged**